storage: TestPebbleMetricEventListener failed

storage.TestPebbleMetricEventListener [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8598302?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8598302?buildTab=artifacts#/) on master @ [622956b9317831449010ba69207dbd112d48f9ab](https://github.com/cockroachdb/cockroach/commits/622956b9317831449010ba69207dbd112d48f9ab):


```
=== RUN   TestPebbleMetricEventListener
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c5eb8fc8b8e683c19f3c3e4238f64094/logTestPebbleMetricEventListener3423442296
    test_log_scope.go:79: use -show-logs to present logs inline
    pebble_test.go:277: -- test log scope end --
    test_log_scope.go:412: unlinkat /artifacts/tmp/_tmp/c5eb8fc8b8e683c19f3c3e4238f64094/logTestPebbleMetricEventListener3423442296: directory not empty
--- FAIL: TestPebbleMetricEventListener (0.01s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPebbleMetricEventListener.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24254

--- a/pkg/storage/pebble.go
+++ b/pkg/storage/pebble.go
@@ -735,6 +735,7 @@ type Pebble struct {
 		syncutil.Mutex
 		flushCompletedCallback func()
 	}
+	asyncDone sync.WaitGroup
 
 	// supportsRangeKeys is 1 if the database supports range keys. It must
 	// be accessed atomically.
@@ -1006,7 +1007,7 @@ func NewPebble(ctx context.Context, cfg PebbleConfig) (p *Pebble, err error) {
 	oldDiskSlow := lel.DiskSlow
 	lel.DiskSlow = func(info pebble.DiskSlowInfo) {
 		// Run oldDiskSlow asynchronously.
-		go oldDiskSlow(info)
+		p.async(func() { oldDiskSlow(info) })
 	}
 	el := pebble.TeeEventListener(
 		p.makeMetricEtcEventListener(ctx),
@@ -1087,6 +1088,17 @@ func NewPebble(ctx context.Context, cfg PebbleConfig) (p *Pebble, err error) {
 	return p, nil
 }
 
+// async launches the provided function in a new goroutine. It uses a wait group
+// to synchronize with (*Pebble).Close to ensure all launched goroutines have
+// exited before Close returns.
+func (p *Pebble) async(fn func()) {
+	p.asyncDone.Add(1)
+	go func() {
+		defer p.asyncDone.Done()
+		fn()
+	}()
+}
+
 func (p *Pebble) makeMetricEtcEventListener(ctx context.Context) pebble.EventListener {
 	return pebble.EventListener{
 		WriteStallBegin: func(info pebble.WriteStallBeginInfo) {
@@ -1132,7 +1144,7 @@ func (p *Pebble) makeMetricEtcEventListener(ctx context.Context) pebble.EventLis
 
 					log.Fatalf(ctx, "file write stall detected: %s", info)
 				} else {
-					go log.Errorf(ctx, "file write stall detected: %s", info)
+					p.async(func() { log.Errorf(ctx, "file write stall detected: %s", info) })
 				}
 				return
 			}
@@ -1171,6 +1183,9 @@ func (p *Pebble) Close() {
 		return
 	}
 	p.closed = true
+
+	// Wait for any asynchronous goroutines to exit.
+	p.asyncDone.Wait()
 
 	handleErr := func(err error) {
 		if err == nil {

